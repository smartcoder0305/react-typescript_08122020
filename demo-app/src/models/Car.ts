export type Car = {
  id: number;
  make: string;
  model: string;
  year: number;
  color: string;
  price: number;
};
