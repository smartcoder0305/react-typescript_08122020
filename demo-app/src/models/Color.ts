

export type Color = {
  id: number,
  name: string,
  hexcode: string,
};