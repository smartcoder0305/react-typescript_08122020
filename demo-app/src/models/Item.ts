export interface Item {
  id?: number;
}