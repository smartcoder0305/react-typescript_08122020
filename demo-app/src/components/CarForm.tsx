import React, { FC } from 'react';

import { Car } from '../models/Car';
import { carColors } from '../carFormLookups';
import { useForm } from '../hooks/useForm';

export type CarFormProps = {
  buttonText: string,
  onSubmitCar: (car: Car) => void,
};

export const CarForm: FC<CarFormProps> = (props) => {

  const [ carForm, change, resetForm ] = useForm({
    make: '', model: '', year: 1900, color: carColors[0], price: 0,
  });

  const submitCar = () => {

    props.onSubmitCar({ ...carForm });

    resetForm();
  };

  return (
    <form>
      <div>
        <label htmlFor="make-input">Make:</label>
        <input type="text" id="make-input" name="make" value={carForm.make} onChange={change} />
      </div>
      <div>
        <label htmlFor="model-input">Model:</label>
        <input type="text" id="model-input" name="model" value={carForm.model} onChange={change} />
      </div>
      <div>
        <label htmlFor="year-input">Year:</label>
        <input type="number" id="year-input" name="year" value={carForm.year} onChange={change} />
      </div>
      <div>
        <label htmlFor="color-select">Color:</label>
        <select id="color-select" name="color" onChange={change}>
          {carColors.map(carColor =>
            <option key={carColor} value={carColor}>{carColor}</option>)}
        </select>
      </div>
      <div>
        <label htmlFor="price-input">Price:</label>
        <input type="number" id="price-input" name="price" value={carForm.price} onChange={change} />
      </div>
      <button type="button" onClick={submitCar}>{props.buttonText}</button>
    </form>

  );

};

CarForm.defaultProps = {
  buttonText: 'Submit Car',
};
