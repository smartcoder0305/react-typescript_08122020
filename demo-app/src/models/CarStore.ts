import { Car } from '../models/Car';
import { CarTableFilterSettings } from '../components/CarTableFilterForm';
import { CarTableSortSettings } from '../components/CarTable';

export type CarStore = {
  cars: Car[],
  editCarId: number,
  carTableSortSettings: CarTableSortSettings,
  confirmDeleteCarId: number,

  setEditCarId: (carId: number) => void,
  confirmDeleteCar: (carId: number) => void,
  deleteCar: (carId: number) => void,
  getCarDetailsById: (carId: number) => string,

  cancelCar: () => void,
  dismissConfirmDeleteCarModal: () => void,
  saveCar: (car: Car) => void,
  addCar: (car: Car) => void,

  setCarTableFilterSettings: (filterSettings: CarTableFilterSettings) => void,
  setCarTableSortSettings: (sortSettings: CarTableSortSettings) => void,
};