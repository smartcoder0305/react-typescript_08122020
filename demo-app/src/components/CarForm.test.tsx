import * as React from 'react';
import {
  render, mount, shallow,
  ReactWrapper, ShallowWrapper,
} from 'enzyme';

import { Car } from '../models/Car';
import { CarForm } from './CarForm';

describe('<CarForm /> Enzyme Static HTML', () => {
  test('<CarForm /> renders', () => {
    const component = JSON.stringify(
      render(<CarForm onSubmitCar={() => null} buttonText="Add Car" />).html(),
    );

    expect(component).toMatchSnapshot();
  });
});

describe('<CarForm /> Enzyme Mock DOM', () => {
  const eventHandlers = {
    submitCar: () => null,
  };

  let submitCarSpy: jest.SpyInstance;
  let component: ReactWrapper;

  beforeEach(() => {
    submitCarSpy = jest.spyOn(eventHandlers, 'submitCar');

    component = mount(
      <CarForm buttonText="Add Car" onSubmitCar={eventHandlers.submitCar} />,
    );
  });

  test('<CarForm /> renders', () => {
    expect(component.find('button').text()).toBe('Add Car');
  });

  test('<CarForm /> form and submit car button', () => {
    const car: Car = {
      make: 'Ford',
      model: 'F-150',
      year: 1980,
      color: 'red',
      price: 42000,
    };

    Object.keys(car).forEach((propName) => {
      const reactWrapper = propName === 'color'
        ? component.find(`#${propName}-select`)
        : component.find(`#${propName}-input`);

      reactWrapper.simulate('change', {
        target: { value: car[propName as keyof Car], name: propName },
      });
    });

    component.find('button').simulate('click');

    expect(submitCarSpy).toHaveBeenCalledWith(car);
  });
});

describe('<CarForm /> Shallow with Enzyme', () => {
  const eventHandlers = {
    submitCar: () => null,
  };

  let submitCarSpy: jest.SpyInstance;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    submitCarSpy = jest.spyOn(eventHandlers, 'submitCar');

    wrapper = shallow(
      <CarForm buttonText="Add Car" onSubmitCar={eventHandlers.submitCar} />,
    );
  });

  test('<CarForm /> renders', () => {
    expect(wrapper.find('button').text()).toBe('Add Car');
  });

  test('<CarForm /> form and submit widget button', () => {
    const car: Car = {
      make: 'Ford',
      model: 'F-150',
      year: 1980,
      color: 'red',
      price: 42000,
    };

    Object.keys(car).forEach((propName) => {
      const shallowWrapper = propName === 'color'
        ? wrapper.find(`#${propName}-select`)
        : wrapper.find(`#${propName}-input`);
      shallowWrapper.simulate('change', {
        target: { value: car[propName as keyof Car], name: propName },
      });
    });

    wrapper.find('button').simulate('click');

    expect(submitCarSpy).toHaveBeenCalledWith(car);
  });
});
