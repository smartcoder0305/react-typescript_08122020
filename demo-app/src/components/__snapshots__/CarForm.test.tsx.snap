// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CarForm /> Enzyme Static HTML <CarForm /> renders 1`] = `"\\"<div><label for=\\\\\\"make-input\\\\\\">Make:</label><input type=\\\\\\"text\\\\\\" id=\\\\\\"make-input\\\\\\" name=\\\\\\"make\\\\\\" value></div><div><label for=\\\\\\"model-input\\\\\\">Model:</label><input type=\\\\\\"text\\\\\\" id=\\\\\\"model-input\\\\\\" name=\\\\\\"model\\\\\\" value></div><div><label for=\\\\\\"year-input\\\\\\">Year:</label><input type=\\\\\\"number\\\\\\" id=\\\\\\"year-input\\\\\\" name=\\\\\\"year\\\\\\" value=\\\\\\"1900\\\\\\"></div><div><label for=\\\\\\"color-select\\\\\\">Color:</label><select id=\\\\\\"color-select\\\\\\" name=\\\\\\"color\\\\\\"><option value=\\\\\\"red\\\\\\">red</option><option value=\\\\\\"green\\\\\\">green</option><option value=\\\\\\"blue\\\\\\">blue</option></select></div><div><label for=\\\\\\"price-input\\\\\\">Price:</label><input type=\\\\\\"number\\\\\\" id=\\\\\\"price-input\\\\\\" name=\\\\\\"price\\\\\\" value=\\\\\\"0\\\\\\"></div><button type=\\\\\\"button\\\\\\">Add Car</button>\\""`;
